core/rawdb/accessors_chain.go

```diff
@@ -522,7 +522,19 @@ func ReadBlock(db ethdb.Reader, hash common.Hash, number uint64) *types.Block {
 	if body == nil {
 		return nil
 	}
-	return types.NewBlockWithHeader(header).WithBody(body.Transactions, body.Uncles).WithExtData(body.Version, body.ExtData)
+	block := types.NewBlockWithHeader(header).WithBody(*body)
+	bodyExtra := types.GetBodyExtra(body)
+
+	extras := types.GetBlockExtra(block)
+	extras.Version = bodyExtra.Version
+	if bodyExtra.ExtData != nil {
+		extras.ExtData = common.PointerTo(make([]byte, len(*bodyExtra.ExtData)))
+		copy(*extras.ExtData, *bodyExtra.ExtData)
+	} else {
+		extras.ExtData = common.PointerTo(make([]byte, 0))
+	}
+
+	return block
```

Comment:
Note:

1. this is about to be thrown away and replaced by hooks in #772 
2. most of the logic comes from `WithExtData` which had to be changed because we cannot set block fields (header field) directly anymore, we need to use NewBlockWithHeader.